rt/transports/jms/src/main/java/org/apache/cxf/transport/jms/JMSDestination.java

```diff
@@ -174,9 +174,9 @@ public void onException(JMSException exception) {
         }
     }
 
-    protected void restartConnection() {
+    protected synchronized void restartConnection() {
```

Comment:
To be fair, I am not sure this is the right thing. Just looking over the code, I think we could run into the deadlock pretty easy. For example, `JMSDestination::activate()` executes `restartConnection()` in `unnamed` thread, whereas the `ExceptionListener` will be executed in the context of `PollingMessageListenerContainer` thread pool. If the the exception happens in one of the pollers while the `restartConnection()` is still in progress (I think it is very possible), we would have an issue.

Reply:
I don't think we're getting a deadlock. The first thread calling `restartConnection()` (`unnamed` or poller-thread) will do the restart - other threads will wait/block until `restartConnection()` is finished. The `ExceptionListener` is only called from the poller threads, these will wait until `restartConnection()` is finished.
So all threads get blocked until the restart ist successful. Without a connection the threads cann't do anything useful - I think this is ok when they are blocked.

One Problem can occur: when `restartConnection()` runs in `createTargetDestinationListener()` to  `container.start()` another `ThreadPool` will be created and the threads are blocked - every time until restart is successful. 
I have no idea how to solve this problem, but I think it's not critical because at this point we already have a successful JMS connection. 

BTW:
When you use the flag  jmsConfig.isOneSessionPerConnection = true, the Exception Handling will not work. The `ExceptionHander `is alwas null and `PollingMessageListenerContainer#handleException()` run into a NullPointerException, so I ignored this case.

I think the interaction between `JMSDestination`, `PollingMessageListenerContainer` and the poller-threads is very complicated and fragile. I think for future versions there must be a rewrite of the code.

Reply:
Sorry for delayed response @steingebein, I think you are right, no risk of deadlock, just tons of blocked threads, did some tests to verify that, we should be "safe" there. The `synchronized` would indeed help there. 

I was trying to reconstruct the flows, and came to the same conclusions as you: the code is quite complicated and fragile.

Reply:
I have one suggestion to improve the code:

Move `sychnronized` from `JMSDestination#restartConnection()` to `PollingMessageListenerContainer#handleException()`, because `PollingMessageListenerContainer` starts multiple threads and needs the synchronisation. So `sychnronized` belongs more into this class.

To complete this change, I have to add `!jmsListener.isRunning() &&` in `JMSDestination#activate()`:
```
  if (!jmsListener.isRunning() && !jmsConfig.isOneSessionPerConnection()) {
                // If first connect fails we will try to establish the connection in the background
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        restartConnection();
                    }
                }).start();
            }
```

But I'm not sure if it makes the code better...

Reply:
I think with the current implementation, it could be an option to consider, but in general, the `ExceptionListener` is a black box for `PollingMessageListenerContainer`, it is actually `JMSDestination` which makes things worst: it tries to reconnect within the callback. Between these two patches, I would personally prefer to stick with your first idea, `synchronized` over `restartConnection`. Thank you.